src/Parallel/ResourceInfo.hpp

```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
+    }
+
+    const auto& singleton_map =
+        tuples::get<LocalTag<Component>>(singleton_map_);
+    return SingletonInfoHolder<Component>{
+        {static_cast<int>(*singleton_map.second)}, singleton_map.first};
+  }
+
+  /// Returns a `std::unordered_set<size_t>` of processors that array components
+  /// should avoid placing elements on. This should be passed to the
+  /// `allocate_array` function of the array component
+  std::unordered_set<size_t> procs_to_ignore() {
+    if (not singleton_map_has_been_set_) {
+      singleton_map_not_built();
+    }
+    return procs_to_ignore_;
+  }
+
+  /// Returns the proc that the singleton `Component` should be placed on.
+  template <typename Component>
+  size_t proc_for() {
+    if (not singleton_map_has_been_set_) {
+      singleton_map_not_built();
+    }
+    return *tuples::get<LocalTag<Component>>(singleton_map_).second;
+  }
+
+  /// \brief Actually builds the singleton map and allocates all the singletons.
+  ///
+  /// \details This could be done in the constructor, however, since we need the
+  /// number of nodes to do some sanity checks, it can't. If an executable is
+  /// run with the --check-options flag, we will be running on 1 proc and 1 node
+  /// so some of the checks done in this function would fail. Unfortunately,
+  /// that means the checks that require knowing the number of nodes now occur
+  /// at runtime instead of option parsing. This is why the
+  /// `singleton_map_has_been_set_` bool is necessary and why we check if this
+  /// function has been called in get_singleton_info(), procs_to_ignore(), and
+  /// proc_for().
+  ///
+  /// This function should only be called once.
+  void build_singleton_map(const Parallel::GlobalCache<Metavariables>& cache);
+
+ private:
+  void singleton_map_not_built() {
+    ERROR(
+        "The singleton map has not been built yet. You must call "
+        "build_singleton_map() before you call this function.");
+  }
+  bool avoid_global_proc_0_{false};
+  bool singleton_map_has_been_set_{false};
+  // These are quantities that we will need for placing singletons which can be
+  // determined just by option parsing
+  size_t num_exclusive_singletons_{};
+  size_t num_procs_to_ignore_{};
+  size_t num_requested_exclusive_singletons_{};
+  size_t num_requested_nonexclusive_singletons_{};
+  std::unordered_multiset<size_t> requested_nonexclusive_procs_{};
+  // Procs that are exclusive. These may or may not be specifically requested
+  std::unordered_set<size_t> procs_to_ignore_{};
+  // For each singleton (whether it has a SingletonInfo or not), maps whether
+  // it's exclusive and what proc it is on.
+  tuples::tagged_tuple_from_typelist<local_tags> singleton_map_{};
+};
+
+template <typename Metavariables>
+void ResourceInfo<Metavariables>::build_singleton_map(
+    const Parallel::GlobalCache<Metavariables>& cache) {
+  const size_t num_procs =
+      static_cast<size_t>(Parallel::number_of_procs(cache));
```

Comment:
Once a decision on #3976 is made, I'll update this PR to use those functions as I do a lot of `static_cast`ing here. Until then, I'll just leave these as is.